DolphinQt: Add support for a Steam Runtime build

OatmealDome opened the pull request:
This PR adds an easy flag to disable/enable things necessary for a Steam build, and makes changes to DolphinQt to create a build that will run in the Steam Runtime.

I built and tested this PR with Steam Runtime version 1, ``scout``, using the revision currently being used in general availability with the Steam client ([which is what Valve recommends for games](https://gitlab.steamos.cloud/steamrt/scout/sdk)).

## Building

Building Dolphin for the Steam Runtime is just like building for a regular Linux system. However, it must must be done within the Steam Runtime environment to prevent system libraries leaking into the application. A chroot environment or Docker container can be used, as outlined [here](https://github.com/ValveSoftware/steam-runtime#building-in-the-runtime).

Some extra CMake flags are required to produce a Steam Runtime build, as some libraries included in the runtime are too old. The recommended cmake configuration command is:

``
cmake .. -G Ninja -DCMAKE_C_COMPILER="gcc-9" -DCMAKE_CXX_COMPILER="g++-9" -DENABLE_EVDEV=OFF -DENABLE_LLVM=OFF -DUSE_SHARED_LIBPNG=OFF -DSTEAM=ON -DENABLE_SDL=ON -DCMAKE_PREFIX_PATH=/path/to/qt
``

## Running

Running a build can occur on any Linux machine. However, it is not possible to run Dolphin within the chroot environment or the Docker container used for building, due to libusb initialization on startup.

First, make sure the scout runtime is installed. This can be done by copying and pasting the following into the address bar

Then, run the build:

``
./steam-runtime/run.sh ./dolphin-emu
``

## Acknowledgements

Thanks to delroth, Techjar, and Warepire for their help getting DolphinQt working.

OatmealDome commented on Source/Core/DolphinQt/CMakeLists.txt:
```diff
@@ -570,6 +572,53 @@ else()
+    COMMAND ${CMAKE_COMMAND} -E copy_if_different "${CMAKE_CURRENT_SOURCE_DIR}/qt.conf" "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/qt.conf"
+  )
+
+  # Mark all data files as sources
```
Fixed, thanks

leoetlino commented:
This is going to massively bloat Externals/ -- currently the Qt submodule already wastes 0.5GB of space on macOS and Linux.

Would it be possible to pull in the Qt libs at build time using CMake rather than a submodule (similar to how MoltenVK is downloaded only when necessary)?

OatmealDome commented:
@leoetlino I think it should be possible. A separate git repository would probably need to be created for this, though (I can set one up on my account for now)

OatmealDome commented:
@leoetlino I had a look at turning the submodule into an ExternalProject, but it seems that the ``git clone`` step is executed only on build time, so finding Qt fails when CMake runs. Not sure there's a workaround for this, but maybe I missed something.

leoetlino commented:
Looks like there are alternatives to ExternalProject if you just want to clone a repo at configure time (rather than at build time): https://stackoverflow.com/a/51311151/1636285

Forgot that CMake would need to find Qt at configure time and so downloading Qt at build time would be too late, my bad.

AdmiralCurtiss commented:
As far as I know, the modern CMake way to do this is [FetchContent](https://cmake.org/cmake/help/latest/module/FetchContent.html).

OatmealDome commented:
@leoetlino @AdmiralCurtiss I had no idea this existed - thanks! I've updated the PR to use ``FetchContent``.

shuffle2 commented:
What is the story with SteamOS and packaging Qt (is it based on debian/ubuntu, idk)? If you are building Qt from source, maybe just start with a Qt6 ?

neobrain commented on Source/Core/DolphinQt/CMakeLists.txt:
```diff
@@ -570,6 +575,53 @@ else()
+  # We also want RPATH, not RUNPATH, so disable the new tags.
+  target_link_options(dolphin-emu
+  PRIVATE
+    -Wl,-z,origin
```
```suggestion
    LINKER:-z,origin
```

OatmealDome commented:
@shuffle2 Not SteamOS, but the Steam Runtime. The Steam Runtime provides a stable environment with common libraries and tools which can be built against. When a native Linux application is downloaded from Steam, the client downloads the Steam Runtime and runs the application in it. It's not distro-specific.

More info here: https://partner.steamgames.com/doc/store/application/platforms/linux and https://github.com/ValveSoftware/steam-runtime#introduction.

The new SteamOS on the Deck is Arch-based, but I think the older ones are Debian or Ubuntu based.

(Also, isn't Dolphin incompatible with Qt6? I have no experience with creating Qt applications)

OatmealDome commented:
@shuffle2 

> OK, but for this discussion it basically means we cannot rely on Steam OS and/or Runtime to provide a recent version of Qt, so we must build it ourselves, right?

Yes, I had to build it myself. The Runtime doesn't ship with a copy of Qt.

> Qt build for Steam could be reused for other *nix targets. Maybe this can be a workaround for distros being so late to ship Qt6.

During testing for this PR, I at one point ran Dolphin outside of the Runtime (with the Runtime-compiled Qt and Manjaro system libraries) and it crashed.

shuffle2 commented:
> During testing for this PR, I at one point ran Dolphin outside of the Runtime (with the Runtime-compiled Qt and Manjaro system libraries) and it crashed.

Sounds funky!

delroth commented:
Similar to my comment on #10516, let's try to avoid hiding too much magic/implicit behavior behind a STEAM_RUNTIME flag. Stuff like "not linking with libpng" should be separate options that we happen to enable on the SteamRT builder by default.

We probably still want some flag to have cmake copy the right files to generate an "app bundle" for SteamRT, but that flag should be additional behavior, not controlling behavior re: dependencies and such. Does this make sense?

neobrain commented on Source/Core/DolphinQt/CMakeLists.txt:
```diff
@@ -570,6 +575,53 @@ else()
+  target_link_options(dolphin-emu
+  PRIVATE
+    LINKER:-z,origin
+    -Wl,--disable-new-dtags
```
```suggestion
    LINKER:--disable-new-dtags
```

degasus commented on Source/Core/DolphinQt/CMakeLists.txt:
```diff
@@ -570,6 +575,53 @@ else()
+
+  add_custom_command(TARGET dolphin-emu POST_BUILD
+    COMMAND ${CMAKE_COMMAND} -E copy_directory "${Qt5_DIR}/../../../plugins/platforms" "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/platforms"
+  )
```
AFAIK they can be merged:
```
  add_custom_command(TARGET dolphin-emu POST_BUILD
    COMMAND ${CMAKE_COMMAND} -E make_directory "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/lib"
    COMMAND cp -P "${Qt5_DIR}/../../*.so*" "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/lib"
    COMMAND ${CMAKE_COMMAND} -E copy_directory "${Qt5_DIR}/../../../plugins/platforms" "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/platforms"
  )
```

OatmealDome replied:
Yep, they can - thanks

OatmealDome commented:
I solved the ``FileSystemTests`` failure in ``cd6f27d1``, but I'm honestly not sure why reshuffling the code like that fixes it.

The problem appears to involve accessing free'd memory in ``FstEntry.name`` during ``SaveFst()``. A valgrind output can be found [here](https://pastebin.com/fm1rQqGz).

OatmealDome commented:
OK, I think this is about ready now. Addressed all comments, cleaned up the commits, and the unit tests no longer fail when running them within the runtime.

I've been uploading builds made with this PR to Steam and they seem to work fine.

JMC47 commented:
@dolphin-emu-bot rebuild

iwubcode commented on Source/Core/DolphinQt/CMakeLists.txt:
```diff
@@ -607,6 +607,12 @@ if (CMAKE_SYSTEM_NAME STREQUAL "Linux" AND STEAM)
     INSTALL_RPATH "\$ORIGIN/lib"
   )
+
+  add_custom_command(TARGET dolphin-emu POST_BUILD
```
Is there a reason this isn't done all the time?  Is Qt managed by the system or do we have a special build like on Windows?

OatmealDome replied:
There is no Qt present in the steamrt, so yes, [we provide our own build](https://github.com/OatmealDome/ext-steam-runtime-qt). It's not in a submodule, though. I believe we will just have it on the builder.

iwubcode commented on Source/Core/DolphinQt/CMakeLists.txt:
```diff
@@ -593,6 +593,22 @@ else()
 endif()
 
+if (CMAKE_SYSTEM_NAME STREQUAL "Linux" AND STEAM)
+  # Set that we want ORIGIN in FLAGS.
```
Could possibly use `BUILD_RPATH_USE_ORIGIN` though that wouldn't allow you to disable the tags...

OatmealDome replied:
I think I tried that earlier and it used RUNPATH?

iwubcode commented on Source/Core/Core/IOS/FS/HostBackend/FS.cpp:
```diff
@@ -552,16 +552,20 @@ ResultCode HostFileSystem::Rename(Uid uid, Gid gid, const std::string& old_path,
   if (it != old_parent->children.end())
   {
-    *new_entry = *it;
+    new_entry->data = it->data;
```
Maybe it doesn't matter but this is different behavior than before because all of the `FstEntry` isn't copied.  It'd be nice to keep the copy constructor behavior.  What was the  ut failure?

OatmealDome replied:
[``data``, ``children``, and ``name`` (which we don't care about because it's the old name) are all the members of this struct](https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/Core/IOS/FS/HostBackend/FS.h#L62).

Any test that attempted using Rename would crash when running it within the steamrt. I posted a valgrind output in a comment above.

JMC47 commented:
Looks like reviews have essentially finished.  I asked a couple of times if anyone else had any qualms and/or concerns, and no one has spoken up.  Let us get the ball rolling.

qurious-pixel commented on Source/Core/DolphinQt/CMakeLists.txt:
```diff
@@ -606,6 +606,47 @@ else()
+  add_custom_command(TARGET dolphin-emu POST_BUILD
+    COMMAND ${CMAKE_COMMAND} -E make_directory "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/lib"
+    COMMAND cp -P "${QT_DIR}/../../*.so*" "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/lib"
+    COMMAND ${CMAKE_COMMAND} -E copy_directory "${QT_DIR}/../../../plugins/platforms" "${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/platforms"
```
There should be a better way to find the plugins path. This is failing on system libs for ubuntu, and rolls-back the compilation. The path for Ubuntu 20.04 for example is: `"${QT_DIR}/../../qt/plugins/platforms"`
And why copy a blank `qt.conf` file. Where is it being set?

delroth replied:
Building with system libs is not expected to work, since by definition that wouldn't be a steam runtime build.

arseniy-sonar replied:
Nicely spotted, see my bug report: https://bugs.dolphin-emu.org/issues/12946

delroth replied:
Should be fixed by #10728, steamrt builds shouldn't have defaulted to ON. Thanks for the reports.
